### PR TITLE
Add GreatCircleLayer

### PR DESCRIPTION
**Pull request by XiaokaiUber**
For #2499 

#### Background
customize ArcLayer to draw a great circle between two points on the map

#### Change List
- implement GreatCircleLayer by customizing ArcLayer
- added vertex shader for GreatCircleLayer
- add GreatCircleLayer example in layer browser

#### result
great-circle
arc

### CODE REVIEW

**Comment by ibgreen** [comment]
> Changes made to ArcLayer are reverted. GreatCircleLayer is created as a subclass of ArcLayer in the experiment layer folder. A GreatCircleLayer example is added to layer browser(Experimental Core Layers) as can be seen below:

Thanks, that makes a lot more sense. 

I could argue that this might be more of an "example specific layer" rather than an "experimental layer" (i.e. users would have to copy it from an example rather than import it from experimental layers), but I am fine with either approach. 

Having it in experimental layers makes sense if we think there is a future layer catalog where this layer could fit in, I could for instance imagine a layer catalog with various geospatial "helper layers", other layers could perhaps a "compass layer", ... 

Also, you could probably build some really cool geospatial grid layer/satellite orbit layers using this layer as a primitive.

File: examples/layer-browser/src/examples/experimental-layers.js [hunk 1]
```diff
@@ -213,6 +228,7 @@ export default {
     AdvancedTextLayer: AdvancedTextLayerExample,
     GPUGridLayer: GPUGridLayerExample,
     'GPUGridLayer (1M)': GPUGridLayerPerfExample('1M', dataSamples.getPoints1M),
-    'GPUGridLayer (5M)': GPUGridLayerPerfExample('5M', dataSamples.getPoints5M)
+    'GPUGridLayer (5M)': GPUGridLayerPerfExample('5M', dataSamples.getPoints5M),
+    'ArcLayer (Great Circle)': ArcLayerGreatCircleExample
```
Comment (Pessimistress) [inline examples/layer-browser/src/examples/experimental-layers.js]: `GreatCircleLayer`

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+vec2 toDegree(vec2 v) {
+  return v * 180.0 / PI;
+}
+
+float getAngularDist (vec2 source, vec2 target) {
+  vec2 delta = source - target;
+  float a = 
+    sin(delta.y * 0.5) * sin(delta.y * 0.5) + 
+    cos(source.y * PI / 180.) * cos(target.y * PI / 180.) *
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: Use `toRadian`

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+vec2 toDegree(vec2 v) {
+  return v * 180.0 / PI;
+}
+
+float getAngularDist (vec2 source, vec2 target) {
+  vec2 delta = source - target;
+  float a = 
+    sin(delta.y * 0.5) * sin(delta.y * 0.5) + 
+    cos(source.y * PI / 180.) * cos(target.y * PI / 180.) * 
+    sin(delta.x / 2.0) * sin(delta.x / 2.);
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: Code style: you have `/ 2.0`, `/ 2.`, `* 0.5`, just pick one

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	return 2.0 * atan(sqrt(a), sqrt(1.0 - a));
+}
+
+vec2 interpolate (vec2 source, vec2 target, float t) {
+	float angularDist = getAngularDist(source, target);
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: This is called twice for the same arguments.

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  float nextSegmentRatio = getSegmentRatio(segmentIndex + indexDir);
+  
+  vec2 source = radians(instancePositions.xy);
+  vec2 target = radians(instancePositions.zw);
+  
+  float angularDist = getAngularDist(source, target);
+
+  vec3 currPos = project_position(vec3(degrees(interpolate(source, target, angularDist, segmentRatio)), 0.0));
+  vec3 nextPos = project_position(vec3(degrees(interpolate(source, target, angularDist, nextSegmentRatio)), 0.0));
+  vec4 curr = project_to_clipspace(vec4(currPos, 1.0));
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: Use `project_position_to_clipspace`

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+float getSegmentRatio(float index) {
+  return smoothstep(0.0, 1.0, index / (numSegments - 1.0));
+}
+
+// get angular distance in radian
+float getAngularDist (vec2 source, vec2 target) {
+  vec2 delta = source - target;
+  float a =
+    sin(delta.y / 2.0) * sin(delta.y / 2.0) +
+    cos(radians(source.y)) * cos(radians(target.y)) *
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: source and target are already in `radians`. Why again?

File: examples/layer-browser/src/data-samples.js [hunk 7]
```diff
@@ -109,6 +109,17 @@ export const zigzag = [
...
+export const greatCircles = [
+  {source: [-20, 0], target: [-160, 0]},
+  {source: [-20, 10], target: [-160, 10]},
+  {source: [-20, 20], target: [-160, 20]},
+  {source: [-20, 30], target: [-160, 30]},
+  {source: [-20, 40], target: [-160, 40]},
+  {source: [-20, 50], target: [-160, 50]},
+  {source: [-20, 60], target: [-160, 60]},
+  {source: [-20, 70], target: [-160, 70]}
+];
```
Comment (Pessimistress) [inline examples/layer-browser/src/data-samples.js]: Need more test cases. E.g. longitude 10 to longitude -170. Source and target not on the same latitude.

**Comment by Pessimistress** [comment]
Please update the PR title and summary to reflect the latest changes.

File: examples/layer-browser/src/data-samples.js (hunk 7 above)
Comment (XiaokaiUber) [reply examples/layer-browser/src/data-samples.js]: more test cases are added.
great_circle_example

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  vec2 source = radians(instancePositions.xy);
+  vec2 target = radians(instancePositions.zw);
+  
+  float angularDist = getAngularDist(source, target);
+
+  vec3 currPos = vec3(degrees(interpolate(source, target, angularDist, segmentRatio)), 0.0);
+  vec3 nextPos = vec3(degrees(interpolate(source, target, angularDist, nextSegmentRatio)), 0.0);
+
+  vec4 curr = project_position_to_clipspace(currPos, instancePositions64Low.xy, vec3(0.0));
+  vec4 next = project_position_to_clipspace(nextPos, instancePositions64Low.zw, vec3(0.0));
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: Although `instancePositions64Low` is not significant in this use case, you need to make sure that the `next` position in one segment and the `curr` position in the following segment match.

File: modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  
+  vec2 source = radians(instancePositions.xy);
+  vec2 target = radians(instancePositions.zw);
+  
+  float angularDist = getAngularDist(source, target);
+
+  vec3 currPos = vec3(degrees(interpolate(source, target, angularDist, segmentRatio)), 0.0);
+  vec3 nextPos = vec3(degrees(interpolate(source, target, angularDist, nextSegmentRatio)), 0.0);
+
+  vec2 currPos64Low = degrees(interpolate(instancePositions64Low.xy, instancePositions64Low.zw, angularDist, segmentRatio));
```
Comment (Pessimistress) [inline modules/experimental-layers/src/great-circle-layer/great-circle-vertex.glsl.js]: - This is mathematically incorrect
- 64bit accuracy of interpolated points is not a concern for this layer

So just use linear interpolation.

**Pull request merged**